content/blog/2020-12-09-slp/index.md

```diff
@@ -0,0 +1,184 @@
+   - To simplify the search space for the next part, we 
+     only allow a statement to occupy the beginning or end instruction in a 
+     pack list once. This way, we don't have to choose between multiple options 
+     later down the line.
+   ![phase2](phase2.png)
+3. Now that we have pairs of instructions that can be executed by a single SIMD
+   instruction without any issues, we can combine multiple of these pairs into 
+   an even wider SIMD instruction. Taking advantage of the last bullet above, 
+   we only combine two lists if the beginning and end instruction is the same.
+   ![phase3](phase3.png)
+4. The last part is to actually generate the concrete SIMD instructions.
+   Since statement groups could depend on data defined by a previous one, we
+   need to generate the instructions in a topological ordering of the data
+   dependency graph. The original instruction order works so the scheduling 
+   phase generates instructions in the order of the original basic block. To 
+   deal with cycles, we simply revert the pack and use regular instructions 
+   instead.
+   ![phase4](phase4.png)
+
+This routine will take in a single basic block and then convert it to a basic
+block with SIMD instructions.
+
+By simply unrolling loops, we can apply this to loops to get a vector parallel 
+version.
+
+## Performance Comparison
+
+The authors chose to use a benchmark suite that consists of multimedia and 
+scientific programs. These process heavy amounts of data, e.x. 
+matrix multiplication or color conversion. Their experimental platform was a 
+Motorola processor with an AltiVec co-processor, i.e. all vector instructions 
+work on the coprocessor and all sequential operations on the main unit. Any 
+movement between the vector registers and sequential registers must pass 
+through memory.
+
+They don't really compare with any other baseline implementations. They mention 
+previous versions of this in their related work section but it would have been 
+nice to see how many opportunities SLP can exploit over the loop transformation 
+version. However, they do show which opportunities come from vectorizable 
+(loop based) opportunities versus just SLP, which still shows large improvement,
+just not as large.
+
+One claim that they put forth (albeit a reasonable one) is that their approach
+is more robust than the previous loop transformation techniques. It would be
+have been good to see this experimentally verified in their chosen benchmarks.
+
+## Limitations
+
+The authors do a good job of highlighting several points that limit the 
+effectiveness of vectorization.
+- At the time, several architectures had different implementations of vector
+  instructions. Additionally, they don't have first party support but use a 
+  side-unit to achieve this.
+- At the time, there was a lack of support for proper packing/unpacking. 
+  Additionally, you can't move values between scalar and vector registers, 
+  forcing you to use memory.
```

Comment:
Could be fun to look into how vector instruction sets have changed. For one thing, they are extremely common now—basically every Intel/AMD x86 CPU out there now has some form of AVX, for instance. And in the ARM world, [Neon](https://en.wikipedia.org/wiki/ARM_architecture#Advanced_SIMD_(Neon)) is basically universal these days too.

Reply:
I left this as a discussion question 😛